pkg/kapis/devops/v1alpha4/util.go

```diff
@@ -34,3 +41,61 @@ func convertPipelineRunsToObject(prs []v1alpha4.PipelineRun) []runtime.Object {
 	}
 	return result
 }
+
+func convertParameters(payload *devops.RunPayload) []v1alpha4.Parameter {
+	if payload == nil {
+		return nil
+	}
+	var parameters []v1alpha4.Parameter
+	for _, parameter := range payload.Parameters {
+		if parameter.Name == "" || parameter.Value == "" {
+			continue
+		}
+		parameters = append(parameters, v1alpha4.Parameter{
+			Name:  parameter.Name,
+			Value: fmt.Sprint(parameter.Value),
+		})
+	}
+	return parameters
+}
+
+func getScm(ps *v1alpha3.PipelineSpec, branch string) (*v1alpha4.SCM, error) {
+	var scm *v1alpha4.SCM
+	if ps.Type == v1alpha3.MultiBranchPipelineType {
+		if branch == "" {
+			return nil, errors.New("missing branch name for running a multi-branch Pipeline")
+		}
+		// TODO validate if the branch dose exist
+		// we can not determine what is reference type here. So we set reference name only for now
+		scm = &v1alpha4.SCM{
+			RefName: branch,
+			RefType: "",
+		}
+	}
+	return scm, nil
+}
+
+func getPipelineRef(pipeline *v1alpha3.Pipeline) *corev1.ObjectReference {
+	return &corev1.ObjectReference{
+		Kind:      pipeline.Kind,
+		Name:      pipeline.GetName(),
+		Namespace: pipeline.GetNamespace(),
+	}
+}
+
+func createPipelineRun(pipeline *v1alpha3.Pipeline, payload *devops.RunPayload, scm *v1alpha4.SCM) *v1alpha4.PipelineRun {
+	controllerRef := metav1.NewControllerRef(pipeline, pipeline.GroupVersionKind())
+	return &v1alpha4.PipelineRun{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName:    pipeline.GetName() + "-runs-",
+			Namespace:       pipeline.GetNamespace(),
+			OwnerReferences: []metav1.OwnerReference{*controllerRef},
+		},
+		Spec: v1alpha4.PipelineRunSpec{
+			PipelineRef:  getPipelineRef(pipeline),
+			PipelineSpec: &pipeline.Spec,
```

Comment:
Please don't set the field `spec.PipelineSpec` in the APIServer. The controller will do that.

Reply:
IMO, we should persistent the `spec.PipelineSpec` when we create a `PipelineRun`. Assume that `Pipeline` is changed immediately after we create `PipelineRun`, but the controller hasn't had time to reconcile the `PipelineRun`. At last, an unexpected `PipelineSpec` might be set unexpectedly.

Reply:
I don't think so. The controller is the real person how to trigger a Pipeline instead of the APIServer.